File: project_quantum/encryption_utils.py
import hashlib
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import padding
import base64
import os

TEAM_NAME = os.getenv("TEAM_NAME", "unknown_team")

def decrypt_and_get_key(ciphertext_b64: str, state_material: str) -> str:
    if TEAM_NAME == "unknown_team":
        return "ERROR: TEAM_NAME missing in .env"
    try:
        key = hashlib.sha256(str(state_material).encode()).digest()
        ct = base64.b64decode(ciphertext_b64)
        iv = ct[:16]
        cipher = Cipher(algorithms.AES(key), modes.CBC(iv), backend=default_backend())
        decryptor = cipher.decryptor()
        padded_pt = decryptor.update(ct[16:]) + decryptor.finalize()
        
        unpadder = padding.PKCS7(128).unpadder()
        flag = unpadder.update(padded_pt) + unpadder.finalize()
        flag_str = flag.decode('utf-8')
        
        if flag_str.startswith("flag_"):
            return hashlib.md5(f"{TEAM_NAME}_{flag_str}".encode()).hexdigest()[:6]
    except Exception:
        pass
    return "INVALID_STATE_OR_DECRYPTION_FAILED"

File: project_chronos/backend/src/main/java/com/chronos/EncryptionUtils.java
package com.chronos;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.MessageDigest;
import java.util.Base64;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class EncryptionUtils {
    public static String decryptAndGetKey(String ciphertextB64, String password) {
        String teamName = "unknown_team";
        try {
            // Assume the backend is run from project_chronos/backend folder, so .env is at ../.env
            List<String> lines = Files.readAllLines(Paths.get("../.env"));
            for (String line : lines) {
                if (line.startsWith("TEAM_NAME=")) {
                    teamName = line.substring(10).trim();
                    break;
                }
            }
        } catch (Exception e) {}
        
        if ("unknown_team".equals(teamName)) {
            return "ERROR: TEAM_NAME missing in ../.env";
        }
        
        try {
            MessageDigest sha = MessageDigest.getInstance("SHA-256");
            byte[] keyBytes = sha.digest(password.getBytes("UTF-8"));
            
            byte[] ctBytes = Base64.getDecoder().decode(ciphertextB64);
            byte[] iv = new byte[16];
            System.arraycopy(ctBytes, 0, iv, 0, 16);
            byte[] encrypted = new byte[ctBytes.length - 16];
            System.arraycopy(ctBytes, 16, encrypted, 0, encrypted.length);
            
            SecretKeySpec secretKey = new SecretKeySpec(keyBytes, "AES");
            IvParameterSpec ivSpec = new IvParameterSpec(iv);
            
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, secretKey, ivSpec);
            
            byte[] decrypted = cipher.doFinal(encrypted);
            String flag = new String(decrypted, "UTF-8");
            
            if (flag.startsWith("flag_")) {
                MessageDigest md5 = MessageDigest.getInstance("MD5");
                md5.update((teamName + "_" + flag).getBytes("UTF-8"));
                byte[] md5Bytes = md5.digest();
                StringBuilder sb = new StringBuilder();
                for (byte b : md5Bytes) sb.append(String.format("%02x", b));
                return sb.toString().substring(0, 6);
            }
        } catch (Exception e) {}
        return "INVALID_STATE_OR_DECRYPTION_FAILED";
    }
}
